### ISSUE TITLE
log format errors

### ISSUE THREAD

**Issue by teejaswin** [allinurl/goaccess#1621]
goaccess nginx.log --log-format='%d\t%t\t%m\t%h\t%^\t%R\t%u\t%b\t%^\t%^\t%v\t%T\t%s\t%^\t%U\t%^' --date-format=%Y-%m-%d --time-format=%T
Parsed 10 lines producing the following errors:

Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.
Token for '%d' specifier is NULL.

Format Errors - Verify your log/date/time format

**Comment by allinurl** [comment]
Please post some lines from your access log.

**Comment by gggeek** [comment]
Hello.

Same error message here.

This is my log line:

"192.168.56.1, 172.19.1.10, 127.0.0.81" 127.0.0.1 - - [30/Jul/2020:09:02:39 +0100] "GET / HTTP/1.1" 302 901 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36" "-" 109488

And parsing command:

goaccess /var/log/apache2/www.access.log --time-format='%H:%M:%S' --date-format='%d/%b/%Y' -a --all-static-files --keep-last=30 --persist -o /tmp/a.html --log-format='~h{," } %^ - %e %^[%d:%t %^] "%r" %s %b "%R" "%u" %^ %L'

Apart from the fact that it is not too clear if the '%d' token mentioned in the error message stands for date or for day, I have the following questions:
- in the default config log formats, I see that the square brackets are preceded by '%^'. What is correct to match (and skip parsing) a left bracket: is it '%^[' or simply '[' ?
- same for a literal '-' in the log
- and how to match (and skip) the token '"something"', ie. wrapped in double quotes?

**Comment by gggeek** [comment]
Answering to my own question(s):

1. it seems that the correct formats in my case are:

time-format='%H:%M:%S %z'
log-format='~h{ ",} %^[%d:%t] "%r" %s %b "%R" "%u" "%^" %L'

2. it seems that a single '%^' token does greedy matching, ie. it will match any chars, including whitespace (but probably not a double quote). That's why '%^[' can be used to match the string '127.0.0.1 - - ['

It would be nice if the parsing tokens were better explained in the docs and default config file...

**Comment by allinurl** [comment]
@gggeek Thanks for the feedback. I can look into those suggestions.

Can you please post a few lines from your access log so I can take a look at what might be the issue? Thanks again.

**Comment by gggeek** [comment]
Hello. Thanks for replying.

I have pasted in my first comment an example of what my log lines look like, as well as a correct log-format string in the 2nd comment (now reformatted for readability).

I did not keep a copy of the log-format strings which I tested unsuccessfully before managing to build the correct one, however I can tell by memory that I struggled with:

- parsing the bit `  127.0.0.1 - - [ ` : I tried to use 3 tokens for that, unsuccessfully. I settled on using a single `%^` to match and discard it all even though at least the 2nd dash might actually be used by goaccess when it is not a dash but a username (we do use basic auth for some parts of the site)

- understanding what is matched by `%^`

- understanding how to match bits like `-` and `"hello"`
